<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>
<body>
    <pre>
Q.1 Write one example explaining how you can write a callback function ?
Answer 1.
function firstFun(name,sum){
    console.log(name,sum);
}
function callBackfun(a,b){
    return a+b;
}
firstFun("Sum of a and b",callBackfun(20,20))
Output:
Sum of a and b 40


Q. 2 "Write callback function to print numbers from 1 to 7, in which 1 should be printed after 1 sec , 2 should be printed after 2 sec, 3 should be printed after 3 sec and so on.
Explain callback hell.
Numbers
1
2
3
4
5
6
7"
Answer 2.
function callB() {
    let count = 1;
    setTimeout(() => {
        console.log(count++);
        setTimeout(() => {
            console.log(count++);
            setTimeout(() => {
                console.log(count++);
                setTimeout(() => {
                    console.log(count++);
                    setTimeout(() => {
                        console.log(count++);
                        setTimeout(() => {
                            console.log(count++);
                            setTimeout(() => {
                                console.log(count++);
                            }, 1000)
                        }, 2000)
                    }, 3000)
                }, 4000)
            }, 5000)
        }, 6000)
    }, 7000)
}
callB()
output:
1
2
3
4
5
6
7


Q.3 "Write promise function to print numbers from 1 to 7, in which 1 should be printed after 1 sec , 2 should be printed after 2 sec, 3 should be printed after 3 sec and so on.
Numbers
1
2
3
4
5
6
7"
Answer 3.
function getMul() {
    for (let i = 0; i <7; i++) {
        setTimeout(()=>{
            console.log( i + 1);
           },i*1000 );
    }
}
getMul();
output: 
1
2
3
4
5
6
7"

Q 4. "Create a promise function accepting an argument, if yes is passed to the function then it should go to resolved state and print Promise Resolved, and if nothing is passed 
then it should go to reject the state and catch the error and print Promise Rejected "
Answer 4.

const promise=new Promise((resolve,recject)=>{
   let count=10;
    if(count===10){
      resolve("Successfully Resolved");
    }
    else{
        reject("Number not same")
    }
})
promise.then((x)=>{
    console.log(x);
})
.catch((x)=>{
    console.log(x);
        setTimeout(()=>{
            console.log("i am second promise");
            resolve();
        },2000)
    })
Output:
Successfully Resolved


Q 5.Create examples to explain callback function
Answer 5.
        A callback function is a function passed into another function as an argument.
        This function is invoked inside the outer function to complete an action.
        Let's take a simple example of how to use callback function

function callbackFunction(name) {
  console.log("Hello " + name);
}
function outerFunction(callback) {
  callback("Mayuri");
}
outerFunction(callbackFunction);
output: Hello Mayuri
Q 6. Create examples to explain callback hell function
Answer 6. Callback Hell is an anti-pattern with multiple nested callbacks which makes code hard to read and debug when dealing with asynchronous logic.


function callB() {
        let count=0
        setTimeout(() => {
            count++
            console.log(count);
            setTimeout(() => {
                count++
                console.log(count);
                setTimeout(() => {
                    count++
            console.log(count);
                    setTimeout(() => {
                        count++
                        console.log(count);
                        setTimeout(() => {
                            count++
                            console.log(count);
                            setTimeout(() => {
                                count++
                                console.log(count);
                                setTimeout(() => {
                                    count++
                                    console.log(count);
                                }, 1000)
                            }, 2000)
                        }, 3000)
                    }, 4000)
                }, 5000)
            }, 6000)
        }, 7000)
    }
    callB()
    output:
    1
    2
    3
    4
    5
    6
    7


Q 7.Create examples to explain promises function
Answer 7. A promise is an object that may produce a single value some time in the future with either a resolved value or a reason that itâ€™s not resolved(for example, network error). It will be in one of the 3 possible states: fulfilled, rejected, or pending.
function callingPromise(){
    return new Promise((resolve,reject)=>{
        setTimeout(()=>{
            reject("i am rejecting the promise")
        },2000)
    })
}
callingPromise()

.then((response) => {
    console.log(response);
  })
  .catch((error)=>{
    console.log("error");
  }).finally(()=>{
    console.log("end of promise");
  })
Output:
error
end of promise


Q 8. Create examples to explain async await function
Answer 8. We use the async keyword with a function to represent that the function is an asynchronous function. The async function returns a promise.
let promise = new Promise(function (resolve, reject) {
    setTimeout(function () {
    resolve('Promise resolved')}, 1000); 
});
async function asyncFunc() {
    try {

        let result = await promise; 

        console.log(result);
    }   
    catch(error) {
        console.log(error);
    }
}

// calling the async function
asyncFunc(); 
Output: 
Promise resolved


 Q 9. Create examples to explain promise.all function
 Answer 9. Promise.all is a promise that takes an array of promises as an input (an iterable), and it gets resolved when all the promises get resolved or any one of them gets rejected.
let promise1 = new Promise((resolve,reject)=>{
    resolve("Welcome");
})

let promise2 = new Promise((resolve,reject)=>{
     resolve("To")
   
})
let promise3 = new Promise((resolve,reject)=>{
     resolve("Prepbytes")    
})
Promise.all([promise1, promise2, promise3])
    .then((result)=>{
        console.log(result);
    })
    .catch((err)=>{
        console.log(err);
    })
Output:
['Welcome', 'To', 'Prepbytes']

    </pre>
    
    <script src="./index.js"></script>
</body>
</html>